/**
 * SPDX-FileCopyrightText: (c) 2026 Liferay, Inc. https://liferay.com
 * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
 */

package com.liferay.one.service;

import com.liferay.one.constants.ClassNameConstants;
import com.liferay.one.exception.LicenseKeyActiveException;
import com.liferay.one.exception.NoSuchLicenseKeyException;
import com.liferay.one.license.LicenseKeyGenerator;
import com.liferay.one.license.LicenseKeyValidator;
import com.liferay.one.model.LicenseKey;
import com.liferay.one.model.SubscriptionEntry;
import com.liferay.petra.string.StringBundler;
import com.liferay.portal.kernel.util.StringUtil;
import com.liferay.portal.kernel.util.Validator;

import java.text.DateFormat;
import java.text.SimpleDateFormat;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.TimeZone;

import org.json.JSONObject;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;
import org.springframework.http.HttpStatus;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClientResponseException;
import org.springframework.web.util.UriComponentsBuilder;

/**
 * @author Amos Fong
 * @author Allen Ziegenfus
 */
@Component
public class LicenseKeyService extends OneBaseService {

	public LicenseKey addLicenseKey(
			String accountName, boolean active, String additionalInfo,
			boolean complimentary, String description, String domains,
			Date expirationDate, String hostName, String ipAddresses,
			String licenseName, String licenseType, int licenseVersion,
			String macAddresses, int maxClusterNodes, long maxConcurrentUsers,
			int maxHttpSessions, int maxServers, long maxUsers, String name,
			String orderId, String owner, String productExternalId,
			String productName, String productVersion, String serverId,
			String sizing, Date startDate)
		throws Exception {

		_licenseKeyValidator.validateMetadata(
			description, licenseType, maxClusterNodes, name, owner,
			productVersion);

		_licenseKeyValidator.validateDates(
			expirationDate, hostName, ipAddresses, licenseType, macAddresses,
			startDate);

		String key = _licenseKeyGenerator.generateKey(
			accountName, licenseName, licenseType, licenseVersion, productName,
			productExternalId, productVersion, owner, maxClusterNodes,
			maxServers, maxHttpSessions, maxConcurrentUsers, maxUsers, sizing,
			description, domains, hostName, ipAddresses, macAddresses, serverId,
			startDate, expirationDate, new Date());

		JSONObject jsonObject = new JSONObject(
		).put(
			"accountName", accountName
		).put(
			"active", active
		).put(
			"additionalInfo", additionalInfo
		).put(
			"complimentary", complimentary
		).put(
			"customExpirationDate", _toISO8601(expirationDate)
		).put(
			"description", description
		).put(
			"domains", domains
		).put(
			"hostName", hostName
		).put(
			"ipAddresses", ipAddresses
		).put(
			"key", key
		).put(
			"licenseName", licenseName
		).put(
			"licenseType", licenseType
		).put(
			"licenseVersion", licenseVersion
		).put(
			"macAddresses", macAddresses
		).put(
			"maxClusterNodes", maxClusterNodes
		).put(
			"maxConcurrentUsers", maxConcurrentUsers
		).put(
			"maxHttpSessions", maxHttpSessions
		).put(
			"maxServers", maxServers
		).put(
			"maxUsers", maxUsers
		).put(
			"name", name
		).put(
			"orderId", orderId
		).put(
			"owner", owner
		).put(
			"productExternalId", productExternalId
		).put(
			"productName", productName
		).put(
			"productVersion", productVersion
		).put(
			"serverId", serverId
		).put(
			"sizing", sizing
		).put(
			"startDate", _toISO8601(startDate)
		);

		String response = post(
			getAuthorization(), jsonObject.toString(),
			UriComponentsBuilder.fromPath(
				"/o/c/licensekeys"
			).build(
			).toUri());

		return new LicenseKey(new JSONObject(response));
	}

	public LicenseKey extendLicenseKey(
			Date expirationDate, long licenseKeyId, Date startDate)
		throws Exception {

		LicenseKey newLicenseKey = _copyLicenseKey(
			expirationDate, getLicenseKey(licenseKeyId), startDate);

		List<SubscriptionEntry> subscriptionEntries =
			_subscriptionEntryService.getSubscriptionEntries(
				StringBundler.concat(
					"(className eq '", ClassNameConstants.LICENSE_KEY,
					"') and (classPK eq ", licenseKeyId, ")"));

		for (SubscriptionEntry subscriptionEntry : subscriptionEntries) {
			_subscriptionEntryService.addSubscriptionEntry(
				null, ClassNameConstants.LICENSE_KEY,
				newLicenseKey.getLicenseKeyId(),
				subscriptionEntry.getCustomUserId());
		}

		return newLicenseKey;
	}

	public List<LicenseKey> getAssetReceiptLicenseLicenseKeys(
			boolean active, boolean complimentary, String orderId)
		throws Exception {

		return getLicenseKeys(
			StringBundler.concat(
				"(active eq ", active, ") and (complimentary eq ",
				complimentary, ") and (orderId eq '",
				_escapeODataString(orderId), "')"));
	}

	public int getAssetReceiptLicenseLicenseKeysCount(
			boolean active, boolean complimentary, String orderId)
		throws Exception {

		return _getCount(
			StringBundler.concat(
				"(active eq ", active, ") and (complimentary eq ",
				complimentary, ") and (orderId eq '",
				_escapeODataString(orderId), "')"));
	}

	public LicenseKey getLicenseKey(Jwt jwt, long licenseKeyId)
		throws Exception {

		try {
			String response = get(
				getAuthorization(jwt),
				UriComponentsBuilder.fromPath(
					"/o/c/licensekeys/{licenseKeyId}"
				).buildAndExpand(
					licenseKeyId
				).toUri());

			return new LicenseKey(new JSONObject(response));
		}
		catch (WebClientResponseException webClientResponseException) {
			int statusCode = webClientResponseException.getStatusCode(
			).value();

			if (statusCode == HttpStatus.NOT_FOUND.value()) {
				throw new NoSuchLicenseKeyException(
					"No license key exists with ID " + licenseKeyId);
			}

			throw webClientResponseException;
		}
	}

	public LicenseKey getLicenseKey(long licenseKeyId) throws Exception {
		String response = get(
			getAuthorization(),
			UriComponentsBuilder.fromPath(
				"/o/c/licensekeys/{id}"
			).buildAndExpand(
				licenseKeyId
			).toUri());

		return new LicenseKey(new JSONObject(response));
	}

	public LicenseKey getLicenseKeyByExternalReferenceCode(
			String externalReferenceCode)
		throws Exception {

		List<LicenseKey> licenseKeys = getLicenseKeys(
			StringBundler.concat(
				"externalReferenceCode eq '",
				_escapeODataString(externalReferenceCode), "'"));

		if (licenseKeys.isEmpty()) {
			throw new NoSuchLicenseKeyException(
				"{externalReferenceCode=" + externalReferenceCode + "}");
		}

		return licenseKeys.get(0);
	}

	public List<LicenseKey> getLicenseKeys(
			boolean active, boolean complimentary, long entitlementId)
		throws Exception {

		return getLicenseKeys(
			StringBundler.concat(
				"(active eq ", active, ") and (complimentary eq ",
				complimentary, ") and (entitlementId eq '", entitlementId,
				"')"));
	}

	public List<LicenseKey> getLicenseKeys(
			boolean active, String orderId, String productExternalId,
			String serverId)
		throws Exception {

		return getLicenseKeys(
			StringBundler.concat(
				"(active eq ", active, ") and (orderId eq '",
				_escapeODataString(orderId), "') and (productExternalId eq '",
				_escapeODataString(productExternalId), "') and (serverId eq '",
				_escapeODataString(serverId), "')"));
	}

	public List<LicenseKey> getLicenseKeys(String filterString)
		throws Exception {

		return getAllItems("/o/c/licensekeys", filterString, LicenseKey::new);
	}

	public List<LicenseKey> getLicenseKeys(
			String productExternalId, String serverId)
		throws Exception {

		return getLicenseKeys(
			StringBundler.concat(
				"(productExternalId eq '",
				_escapeODataString(productExternalId), "') and (serverId eq '",
				_escapeODataString(serverId), "')"));
	}

	public List<LicenseKey> getLicenseKeys(
			String domains, String licenseType, String owner)
		throws Exception {

		return getLicenseKeys(
			StringBundler.concat(
				"(domains eq '", _escapeODataString(domains),
				"') and (licenseType eq '", _escapeODataString(licenseType),
				"') and (owner eq '", _escapeODataString(owner), "')"));
	}

	public List<LicenseKey> getLicenseKeysByName(
			boolean active, String productName, String serverId)
		throws Exception {

		return getLicenseKeys(
			StringBundler.concat(
				"(active eq ", active, ") and (productName eq '",
				_escapeODataString(productName), "') and (serverId eq '",
				_escapeODataString(serverId), "')"));
	}

	public LicenseKey replaceLicenseKey(
			Date expirationDate, long licenseKeyId, Date startDate)
		throws Exception {

		LicenseKey licenseKey = getLicenseKey(licenseKeyId);

		if (Validator.isNotNull(licenseKey.getOrderId()) &&
			!licenseKey.isActive()) {

			throw new LicenseKeyActiveException();
		}

		LicenseKey newLicenseKey = _copyLicenseKey(
			expirationDate, licenseKey, startDate);

		updateLicenseKey(false, licenseKey.isComplimentary(), licenseKeyId);

		return newLicenseKey;
	}

	public List<LicenseKey> search(
			Boolean active, String description, String hostName,
			String ipAddress, String licenseType, String macAddress,
			String owner, String productExternalId, String productName,
			String serverId)
		throws Exception {

		return getLicenseKeys(
			_buildSearchFilter(
				active, description, hostName, ipAddress, licenseType,
				macAddress, owner, productExternalId, productName, serverId));
	}

	public int searchCount(
			Boolean active, String description, String hostName,
			String ipAddress, String licenseType, String macAddress,
			String owner, String productExternalId, String productName,
			String serverId)
		throws Exception {

		return _getCount(
			_buildSearchFilter(
				active, description, hostName, ipAddress, licenseType,
				macAddress, owner, productExternalId, productName, serverId));
	}

	public LicenseKey updateLicenseKey(
			boolean active, boolean complimentary, long licenseKeyId)
		throws Exception {

		JSONObject jsonObject = new JSONObject(
		).put(
			"active", active
		).put(
			"complimentary", complimentary
		);

		String response = patch(
			getAuthorization(), jsonObject.toString(),
			UriComponentsBuilder.fromPath(
				"/o/c/licensekeys/{id}"
			).buildAndExpand(
				licenseKeyId
			).toUri());

		return new LicenseKey(new JSONObject(response));
	}

	public LicenseKey updateLicenseKeyActive(boolean active, long licenseKeyId)
		throws Exception {

		JSONObject jsonObject = new JSONObject(
		).put(
			"active", active
		);

		String response = patch(
			getAuthorization(), jsonObject.toString(),
			UriComponentsBuilder.fromPath(
				"/o/c/licensekeys/{id}"
			).buildAndExpand(
				licenseKeyId
			).toUri());

		return new LicenseKey(new JSONObject(response));
	}

	private String _buildSearchFilter(
		Boolean active, String description, String hostName, String ipAddress,
		String licenseType, String macAddress, String owner,
		String productExternalId, String productName, String serverId) {

		List<String> conditions = new ArrayList<>();

		if (active != null) {
			conditions.add("(active eq " + active + ")");
		}

		if (Validator.isNotNull(description)) {
			conditions.add(
				"(description eq '" + _escapeODataString(description) + "')");
		}

		if (Validator.isNotNull(hostName)) {
			conditions.add(
				"(hostName eq '" + _escapeODataString(hostName) + "')");
		}

		if (Validator.isNotNull(ipAddress)) {
			conditions.add(
				"(ipAddresses eq '" + _escapeODataString(ipAddress) + "')");
		}

		if (Validator.isNotNull(licenseType)) {
			conditions.add(
				"(licenseType eq '" + _escapeODataString(licenseType) + "')");
		}

		if (Validator.isNotNull(macAddress)) {
			conditions.add(
				"(macAddresses eq '" + _escapeODataString(macAddress) + "')");
		}

		if (Validator.isNotNull(owner)) {
			conditions.add("(owner eq '" + _escapeODataString(owner) + "')");
		}

		if (Validator.isNotNull(productExternalId)) {
			conditions.add(
				"(productExternalId eq '" +
					_escapeODataString(productExternalId) + "')");
		}

		if (Validator.isNotNull(productName)) {
			conditions.add(
				"(productName eq '" + _escapeODataString(productName) + "')");
		}

		if (Validator.isNotNull(serverId)) {
			conditions.add(
				"(serverId eq '" + _escapeODataString(serverId) + "')");
		}

		if (conditions.isEmpty()) {
			return null;
		}

		return String.join(" and ", conditions);
	}

	private LicenseKey _copyLicenseKey(
			Date expirationDate, LicenseKey licenseKey, Date startDate)
		throws Exception {

		return addLicenseKey(
			licenseKey.getAccountName(), true, licenseKey.getAdditionalInfo(),
			licenseKey.isComplimentary(), licenseKey.getDescription(),
			licenseKey.getDomains(), expirationDate, licenseKey.getHostName(),
			licenseKey.getIpAddresses(), licenseKey.getLicenseName(),
			licenseKey.getLicenseType(), licenseKey.getLicenseVersion(),
			licenseKey.getMacAddresses(), licenseKey.getMaxClusterNodes(),
			licenseKey.getMaxConcurrentUsers(), licenseKey.getMaxHttpSessions(),
			licenseKey.getMaxServers(), licenseKey.getMaxUsers(),
			licenseKey.getName(), licenseKey.getOrderId(),
			licenseKey.getOwner(), licenseKey.getProductExternalId(),
			licenseKey.getProductName(), licenseKey.getProductVersion(),
			licenseKey.getServerId(), licenseKey.getSizing(), startDate);
	}

	private String _escapeODataString(String value) {
		if (value == null) {
			return null;
		}

		return StringUtil.replace(value, '\'', "''");
	}

	private int _getCount(String filterString) throws Exception {
		UriComponentsBuilder uriComponentsBuilder =
			UriComponentsBuilder.fromPath(
				"/o/c/licensekeys"
			).queryParam(
				"pageSize", 1
			);

		if (filterString != null) {
			uriComponentsBuilder.queryParam("filter", filterString);
		}

		String response = get(
			getAuthorization(),
			uriComponentsBuilder.build(
			).toUri());

		JSONObject jsonObject = new JSONObject(response);

		return jsonObject.optInt("totalCount");
	}

	private String _toISO8601(Date date) {
		if (date == null) {
			return null;
		}

		DateFormat dateFormat = new SimpleDateFormat(
			"yyyy-MM-dd'T'HH:mm:ss'Z'");

		dateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));

		return dateFormat.format(date);
	}

	@Autowired
	private LicenseKeyGenerator _licenseKeyGenerator;

	@Autowired
	private LicenseKeyValidator _licenseKeyValidator;

	@Autowired
	@Lazy
	private SubscriptionEntryService _subscriptionEntryService;

}